File: src/test/java/uk/gov/justice/hmpps/casenotes/services/EventListenerTest.java
package uk.gov.justice.hmpps.casenotes.services;

import com.google.gson.GsonBuilder;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;
import wiremock.org.apache.commons.io.IOUtils;

import java.io.IOException;

import static java.nio.charset.StandardCharsets.UTF_8;
import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
public class EventListenerTest {

    @Mock
    private CaseNoteService caseNoteService;

    @Mock
    private MergeOffenderService mergeOffenderService;

    private EventListener eventListener;

    @BeforeEach
    public void setup() {
        eventListener = new EventListener(caseNoteService, mergeOffenderService, new GsonBuilder().create());
    }

    @Test
    public void testDeleteEvent() throws IOException {
        when(caseNoteService.deleteCaseNotesForOffender(eq("A1234AA"))).thenReturn(3);

        eventListener.handleEvents(getJson("offender-deletion-request.json"));

        verify(caseNoteService).deleteCaseNotesForOffender(eq("A1234AA"));
    }


    @Test
    public void testMergeEvent() throws IOException {
        when(mergeOffenderService.checkAndMerge(eq(100001L))).thenReturn(2);

        eventListener.handleEvents(getJson("booking-number-changed.json"));

        verify(mergeOffenderService).checkAndMerge(eq(100001L));
    }

    private String getJson(final String filename) throws IOException {
        return IOUtils.toString(getClass().getResourceAsStream(filename), UTF_8.toString());
    }
}

File: src/test/java/uk/gov/justice/hmpps/casenotes/services/EventListenerTest.kt
package uk.gov.justice.hmpps.casenotes.services

import com.fasterxml.jackson.databind.ObjectMapper
import org.junit.jupiter.api.Test
import org.mockito.Mockito.mock
import org.mockito.Mockito.verify
import org.mockito.kotlin.any
import org.mockito.kotlin.whenever
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.json.JsonTest
import wiremock.org.apache.commons.io.IOUtils
import java.nio.charset.StandardCharsets

@JsonTest
class EventListenerTest(@Autowired objectMapper: ObjectMapper) {
  private val mergeOffenderService: MergeOffenderService = mock()
  private var eventListener: EventListener = EventListener(mergeOffenderService, objectMapper)

  @Test
  fun testMergeEvent() {
    whenever(mergeOffenderService.checkAndMerge(any())).thenReturn(2)
    eventListener.handleEvents(getJson("booking-number-changed.json"))
    verify(mergeOffenderService).checkAndMerge(100001L)
  }

  private fun getJson(filename: String): String =
    IOUtils.toString(this::class.java.getResourceAsStream(filename), StandardCharsets.UTF_8.toString())
}
